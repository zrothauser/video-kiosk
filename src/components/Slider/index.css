.b-slider {
  border: 0 none;
  position: relative;
  overflow: visible;
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
}

.b-slider__track {
  border-radius: 5px;
  position: absolute;
  height: 5px;
  width: 100%;
  background-color: rgba(255, 255, 255, .5);
  cursor: pointer;
  -webkit-tap-highlight-color: rgba(0,0,0,0);
  -webkit-tap-highlight-color: transparent;
}

.b-slider__progress {
  border-radius: 5px;
  position: absolute;
  height: 5px;
  left: 0;
  background-color: white;
  cursor: pointer;
  -webkit-tap-highlight-color: rgba(0,0,0,0);
  -webkit-tap-highlight-color: transparent;
}

.b-slider__thumb {
  width: 21px;
  height: 21px;
  border-radius: 50%;
  background-size: 21px 21px;
  background-position: 0px 0px;
  background-color: white;
  position: absolute;
  padding: 0;
  margin: 0;
  text-align: center;
  pointer-events: none;
  transform: translateX(-50%);
  cursor: pointer;
  -webkit-tap-highlight-color: rgba(0,0,0,0);
  -webkit-tap-highlight-color: transparent;
}
