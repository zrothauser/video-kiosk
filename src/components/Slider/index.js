// Dependencies
import React from 'react';
import PropTypes from 'prop-types';

// Styles
import './index.css';

/**
 * Generic slider class that can be styled and used
 * anywhere we need a range input or scrubber.
 *
 * This will be used as both the video's progress bar and
 * the audio controls.
 */
class Slider extends React.Component {
  constructor(props) {
    super(props);

    // Binding
    this.handleMouseDown = this.handleMouseDown.bind(this);
    this.handleDrag = this.handleDrag.bind(this);
    this.handleMouseUp = this.handleMouseUp.bind(this);

    // Initial state
    this.state = {
      currentValue: props.value,
      isMouseDown: false,
    };
  }

  /**
   * If the video time has changed, reset our internal state.
   */
  componentWillReceiveProps(nextProps) {
    // Return if nothing changed
    if (nextProps.value === this.state.currentValue) {
      return;
    }

    // Don't update the value if we're dragging
    if (this.state.isMouseDown) {
      return;
    }

    this.setState({ currentValue: nextProps.value });
  }

  /**
   * Lets us track if the mouse is down or up, so that we can scrub the slider.
   *
   * @param {Event} event Mousedown event.
   */
  handleMouseDown(event) {
    console.log('handleMouseDown');

    this.setState({ isMouseDown: true });

    // Pass to handleDrag(), so that we can go
    // ahead and update the position
    this.handleDrag(event);

    // Watch for the mouse move, until the dragging is done
    document.addEventListener('mousemove', this.handleDrag);
    document.addEventListener('touchmove', this.handleDrag);
    document.addEventListener('mouseup', this.handleMouseUp);
    document.addEventListener('touchend', this.handleMouseUp);
  }

  /**
   * Update state based on where we're dragging.
   */
  handleDrag(event) {
    const {
      minimum,
      maximum,
    } = this.props;

    console.log('handleDrag');

    if (!this.trackElement) {
      return;
    }

    // Get the progress bar dimensions, and then find
    // where it starts and its width
    const progressBarClientRect = this.trackElement.getBoundingClientRect();
    const progressBarStart = progressBarClientRect.left;
    const progressBarWidth = progressBarClientRect.width;

    // Find the cursor/touch point position: we get different data from
    // touch events than from mouse events
    const mouseX = (event.changedTouches && event.changedTouches.length) ?
      event.changedTouches[0].pageX :
      event.pageX;

    // Calculate distance from the beginning, and get a percent of the total
    // width from that. Then convert to a value based on the minimum and maxium
    // possible values.
    const distanceFromBeginning = mouseX - progressBarStart;
    const percent = (distanceFromBeginning / progressBarWidth);
    let newValue = percent * (maximum - minimum);

    // Rounding based on the cursor position could cause
    // the value to go over or under the min/max, so don't
    // let that happen
    if (newValue > maximum) {
      newValue = maximum;
    } else if (newValue < minimum) {
      newValue = minimum;
    }

    console.log('newValue: ' + newValue);

    this.setState({ currentValue: newValue });
  }

  /**
   * Lets us track if the mouse is down or up, so that we can scrub the slider.
   *
   * @param {Event} event      Could be a mouseover, mousedown, or change event.
   */
  handleMouseUp() {
    const { handleSeek } = this.props;
    const { currentValue } = this.state;

    this.setState({ isMouseDown: false });

    // Remove event handlers
    document.removeEventListener('mousemove', this.handleDrag);
    document.removeEventListener('touchmove', this.handleDrag);
    document.removeEventListener('mouseup', this.handleMouseUp);
    document.removeEventListener('touchend', this.handleMouseUp);

    // Call the handleSeek prop, this will probably actually
    // do something with the data now that it's done being set
    console.log('about to handleSeek: ' + currentValue);
    handleSeek(currentValue);
  }

  render() {
    const {
      minimum,
      maximum,
    } = this.props;

    const { currentValue } = this.state;

    const valuePercent = ((currentValue - minimum) / (maximum - minimum)) * 100;

    return (
      <div className="b-slider">
        <div
          className="b-slider__track"
          ref={(trackElement) => { this.trackElement = trackElement; }}
          onMouseDown={event => this.handleMouseDown(event)}
          onTouchStart={event => this.handleMouseDown(event)}
        >
          <div
            className="b-slider__progress"
            style={{ width: `${valuePercent}%` }}
          />
        </div>

        <div
          className="b-slider__thumb"
          style={{ left: `${valuePercent}%` }}
          role="slider"
          aria-valuemax={maximum}
          aria-valuemin={minimum}
          aria-valuenow={currentValue}
          tabIndex="0"
        />
      </div>
    );
  }
}

Slider.propTypes = {
  minimum: PropTypes.number.isRequired,
  maximum: PropTypes.number.isRequired,
  value: PropTypes.number.isRequired,
  handleSeek: PropTypes.func.isRequired,
};

export default Slider;
