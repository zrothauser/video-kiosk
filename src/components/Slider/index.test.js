import React from 'react';
import renderer from 'react-test-renderer';
import Slider from './index';

describe('components', () => {
  describe('Slider', () => {
    const mockProps = {
      minimum: 0,
      maximum: 200,
      value: 40,
      handleSeek: jest.fn(),
    };

    it('renders correctly', () => {
      const SliderComponent = renderer.create(
        <Slider {...mockProps} />,
      );

      expect(SliderComponent).toMatchSnapshot();
    });
  });
});
