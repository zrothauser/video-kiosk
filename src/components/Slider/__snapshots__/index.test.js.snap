// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Slider renders correctly 1`] = `
<div
  className="b-slider"
>
  <div
    className="b-slider__track"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="b-slider__progress"
      style={
        Object {
          "width": "20%",
        }
      }
    />
  </div>
  <div
    aria-valuemax={200}
    aria-valuemin={0}
    aria-valuenow={40}
    className="b-slider__thumb"
    role="slider"
    style={
      Object {
        "left": "20%",
      }
    }
    tabIndex="0"
  />
</div>
`;
